src/lf_apply.rs

```diff
@@ -715,69 +704,75 @@ pub(crate) unsafe fn rav1d_loopfilter_sbrow_cols<BD: BitDepth>(
+    level_ptr = &(*f).lf.level[((*f).b4_stride * (sby * sbsz >> ss_ver) as isize) as usize..];
     uv_off = 0 as c_int as ptrdiff_t;
     have_left = 0 as c_int;
-    x = 0 as c_int;
-    while x < (*f).sb128w {
+    for (x, level_ptr) in (0..(*f).sb128w).zip(level_ptr.chunks((32 >> ss_hor) as usize)) {
         filter_plane_cols_uv::<BD>(
             f,
             have_left,
-            level_ptr as *const [u8; 4],
+            level_ptr,
             (*f).b4_stride,
             ((*lflvl.offset(x as isize)).filter_uv[0]).as_mut_ptr() as *const [[u16; 2]; 2],
-            &mut *(*p.offset(1)).offset(uv_off as isize),
-            &mut *(*p.offset(2)).offset(uv_off as isize),
+            &mut *p[1].offset(uv_off as isize),
+            &mut *p[2].offset(uv_off as isize),
             (*f).cur.stride[1],
             cmp::min(32 as c_int, (*f).w4 - x * 32) + ss_hor >> ss_hor,
             starty4 >> ss_ver,
             uv_endy4 as c_int,
             ss_ver,
         );
-        x += 1;
         have_left = 1 as c_int;
         uv_off += 128 >> ss_hor;
-        level_ptr = level_ptr.offset((32 >> ss_hor) as isize);
     }
 }
 
+/// Gives the same starting positions as `chunks`,
+/// but gives the whole remaining slice at that position
+fn suffixes<T>(slice: &[T], n: usize) -> impl Iterator<Item = &[T]> {
+    let mut offset = 0;
+
+    std::iter::from_fn(move || {
+        let new = slice.get(offset..)?;
+        if new.is_empty() {
+            return None;
+        }
+        offset = offset.saturating_add(n);
+        Some(new)
+    })
+}
+
 pub(crate) unsafe fn rav1d_loopfilter_sbrow_rows<BD: BitDepth>(
     f: *const Rav1dFrameContext,
-    p: *const *mut BD::Pixel,
+    p: &[*mut BD::Pixel; 3],
     lflvl: *mut Av1Filter,
     sby: c_int,
 ) {
-    let mut x;
     let have_top = (sby > 0) as c_int;
     let seq_hdr = &***(*f).seq_hdr.as_ref().unwrap();
     let is_sb64 = (seq_hdr.sb128 == 0) as c_int;
     let starty4 = (sby & is_sb64) << 4;
     let sbsz = 32 >> is_sb64;
-    let ss_ver =
-        ((*f).cur.p.layout as c_uint == Rav1dPixelLayout::I420 as c_int as c_uint) as c_int;
-    let ss_hor =
-        ((*f).cur.p.layout as c_uint != Rav1dPixelLayout::I444 as c_int as c_uint) as c_int;
+    let ss_ver = ((*f).cur.p.layout == Rav1dPixelLayout::I420) as c_int;
+    let ss_hor = ((*f).cur.p.layout != Rav1dPixelLayout::I444) as c_int;
     let endy4: c_uint = (starty4 + cmp::min((*f).h4 - sby * sbsz, sbsz)) as c_uint;
     let uv_endy4: c_uint = endy4.wrapping_add(ss_ver as c_uint) >> ss_ver;
     let mut ptr: *mut BD::Pixel;
-    let mut level_ptr: *mut [u8; 4] =
-        ((*f).lf.level).offset((*f).b4_stride * sby as isize * sbsz as isize);
-    ptr = *p.offset(0);
-    x = 0 as c_int;
-    while x < (*f).sb128w {
+    let mut level_ptr: &[[u8; 4]] =
+        &(*f).lf.level[((*f).b4_stride * sby as isize * sbsz as isize) as usize..];
+    ptr = p[0];
+    for (x, level_ptr) in (0..(*f).sb128w).zip(suffixes(level_ptr, 32)) {
```

Comment:
I think this can be done more simply without `suffixes`.  Can you just do `&level_ptr[32 * x as usize..]` with no `zip` or `suffixes`?

Reply:
And the same for the other `suffixes` call.

Reply:
Or you can do `level_ptr = &level_ptr[32..]`.